File: MyMonitorHub.JavaAgent/src/main/java/mymonitorhub/agent/core/EventQueue.java
package mymonitorhub.agent.core;

import mymonitorhub.agent.model.Event;

import java.util.ArrayList;
import java.util.List;

/**
 * Thread-safe event queue. Draining swaps out the list atomically.
 * Mirrors C# EventQueue.
 */
class EventQueue {

    private List<Event> eventList = new ArrayList<Event>();

    /** Atomically drains and returns all queued events, leaving an empty queue. */
    synchronized List<Event> drain() {
        List<Event> result = eventList;
        eventList = new ArrayList<Event>();
        return result;
    }

    synchronized void add(Event event) {
        eventList.add(event);
    }
}

File: MyMonitorHub.Agent/MyMonitorHub.Agent.Core/Core/EventQueue.cs
using System.Collections.Generic;
using MyMonitorHub.Common.WebApi;

namespace MyMonitorHub.Agent.Core
{
    internal sealed class EventQueue
    {
        private List<Event> _eventList = new List<Event>();

        internal List<Event> EventList
        {
            get
            {
                lock (this)
                {
                    var resultList = _eventList;
                    _eventList = new List<Event>();
                    return resultList;
                }
            }
        }

        internal void Add(Event evt)
        {
            lock (this)
            {
                _eventList.Add(evt);
            }
        }
    }
}